### ISSUE TITLE
research: Set up OpenScene and make use of some open dataset on Scene

### ISSUE THREAD

**Issue by tikikun** [menloresearch/voxel-representation#2]
We need to ground the voxel representation ability in a scene based scenario, setting up room scene

Resources:
 - https://github.com/pengsongyou/openscene/
 - https://github.com/Hoyyyaard/LSceneLLM

**Comment by bachvudinh** [comment]
## Research on the existed in-house room data:
- [Replica Dataset](https://github.com/facebookresearch/Replica-Dataset/blob/main/README.md): The Replica Dataset is a dataset of high quality reconstructions of a variety of indoor spaces from Facebook-Research.
 Yesterday, I attempted to download and visualize the data. I also implemented code to voxelize objects within the scene but encountered two issues:
  
   ![Image](https://github.com/user-attachments/assets/3c30b9b5-a9db-45f1-aa63-0c9cf458986a)

  + When using a voxel_size of 0.03, the code execution time became excessive, potentially entering an infinite loop.
  + Set the voxel_size: 0.5 instantly get the weird result. At first, i thought: mesh.ply file contains the geometry for everything in the room (walls, floor, ceiling, sofa, table, etc.) all mixed together. The computer doesn't automatically know which part is a "wall" and which is a "sofa" just by looking at the geometry. So trying to use the semantic files to identify and throw away the wall/floor/ceiling parts of the mesh before we voxelize, so only the objects remain --> but dont result in unexpected data.
  
   ![Image](https://github.com/user-attachments/assets/7656ef0d-0f48-4371-ba56-e39df13e9bef)
   
  Voxelized data after removing wall:
  
  ![Image](https://github.com/user-attachments/assets/5b3af123-2435-40fb-a168-cc0b932faf4a)

- [ScanNet](https://github.com/ScanNet/ScanNet): an RGB-D video dataset containing 2.5 million views in more than 1500 scans, annotated with 3D camera poses, surface reconstructions, and instance-level semantic segmentations.
- [Matterport3D](https://niessner.github.io/Matterport/): a large-scale RGB-D dataset containing 10,800 panoramic views from 194,400 RGB-D images of 90 building-scale scenes.
- nuScenes

## So Today i will 
- Read and based on OpenScene code to voxelize the scene: https://github.com/pengsongyou/openscene/blob/main/dataset/voxelizer.py
- Yesterday I came across and read thru 2 paper that i found insightful with one is based on the other. So i want to spend time totday to deepdive into their paper(detail will be noted in this issue) and their codebase:
   + 3D-LLm: https://arxiv.org/abs/2307.12981
   + Situational Awareness Matters in 3D Vision Language Reasoning: https://arxiv.org/pdf/2406.07544.

**Comment by bachvudinh** [comment]
## Note on the Situational Awareness paper:

![Image](https://github.com/user-attachments/assets/751aa9ca-1bbd-41cf-8a1d-ad0d9e992675)

**Voxelization Mechanism (Section 4.1 - Visual Tokenization)**
The paper critiques prior methods using object detectors (like VoteNet) for visual tokenization, noting they ignore non-object context, lose global scene structure, and lack zero-shot capabilities. Instead, SIG3D adopts a voxel-based approach using the pretrained OpenScene [51] model:
1. **Input:** A raw 3D point cloud p representing the scene.
2. **Discretization (Voxelization):** The continuous 3D space occupied by the point cloud p is discretized into a regular grid of small, fixed-size 3D voxels (experiments suggest 0.02m size is optimal, matching OpenScene's pretraining). This process is denoted V(p). Points falling within the same voxel are implicitly grouped.
3. **Sparse Feature Extraction:** The resulting sparse voxel grid is fed into E3D, which is specifically a **Minkowski Sparse 3D Convolutional Network** [12] (MinkUNet18A architecture). Sparse CNNs are computationally efficient for 3D data as they perform convolutions only on non-empty voxels, leveraging the inherent sparsity of 3D scenes. E3D is the OpenScene encoder, pretrained via distilling CLIP embeddings from rendered 2D views, providing it with rich, open-vocabulary visual understanding capabilities aligned with language.
4. **Feature Selection:** The model takes feature embeddings from an **upsampled bottleneck layer** within the E3D encoder. This provides a balance between spatial resolution and semantic abstraction.
5. **Projection to Bird's-Eye-View (BEV):** The 3D voxel features from the selected layer are projected onto a 2D plane by computing the **mean average** of features along the vertical (z) axis. This creates a 2D feature map representing a top-down or bird's-eye view of the scene.
6. **Token Definition:** Each grid cell in this resulting 2D BEV feature map is treated as a single **visual token** (z3D). The total number of these tokens (Nv) is fixed (e.g., 256 in experiments). Each token thus represents a specific spatial region in the x-y plane, carrying aggregated semantic features from the 3D space above it. Learnable 3D positional encodings (derived from the center coordinates of the original voxels before projection) are added to these tokens later (Section 4.2).

**Comment by bachvudinh** [comment]
## Previously, i finish the pipeline: Load point cloud -> voxel and plotting based on the OpenScene codebase.
- Point Cloud data:

  ![Image](https://github.com/user-attachments/assets/eac193cd-3de5-46e6-af98-c3f564bda950)

- The voxelized data:

  ![Image](https://github.com/user-attachments/assets/f08790de-6508-4b3d-b837-ac0e0d63336f)

## Today, I will:
+ Working on pipeline: from voxel -> embedding using PointNet(will discuss with team to decide this also).
+ Prepare 2D Scan data.

**Comment by bachvudinh** [comment]
## Previously: 
- Finish the pipeline to prepare the 2D Scan Video for Qwen2.5-VL. The Input will be something like this:

  https://github.com/user-attachments/assets/cac6ca68-ac9e-4d0e-a416-ebd6a9f96394
  
  ![Image](https://github.com/user-attachments/assets/abe5a5d2-a5da-42d1-a20c-624fe89b567a)

- BBox Label for our data: Initially downloaded an incorrect segmentation subset lacking 3D object detection data. Located an alternative solution through a preprocessed dataset from the UniDet3D repository on GitHub, which is already available on Hugging Face for immediate use. Will complete and submit the required Terms of Use agreement afterward, rather than waiting several days for data access approval.

## Today:
- Pair with @new5558 : Finalize the dataset and start the training with new data.

**Comment by bachvudinh** [comment]
## Optimized data presentation: 
- Fine-tuned object/background color contrast and adjusted voxel size parameters to maximize model visibility
- Resolved point cloud alignment issue with ScanNet data: https://github.com/manycore-research/SpatialLM/blob/main/EXAMPLE.md#2-align-the-output-point-cloud.

Scannet Training Scene0000_00.bin:

Pixel data with boundingbox: 
![Image](https://github.com/user-attachments/assets/63850ece-d95b-43d5-a5d4-0c8b22f6682e)

Voxelized data with bounding box:
![Image](https://github.com/user-attachments/assets/d6ddcb48-ffe5-4abd-85e8-00aee411f87f)

2D Scan Video:

https://github.com/user-attachments/assets/f9aea18e-f41e-4677-a11e-e1de93501c1b

  Meta Data:
  ```
{
  "lidar_path": "train_0.bin",
  "dataset_name": "multiscan",
  "path_to_video": "2d_scan_aligned/multiscan/train_0.mp4",
  "video_metadata": {
    "pixels_per_unit": 50,
    "frame_wh": [
      224,
      196
    ],
    "z_min": -1.284,
    "z_max": 1.296,
    "scan_steps": 30
  },
  "objects": [
    {
      "object": "door",
      "x": 2.121,
      "y": -0.922,
      "z": -0.218,
      "x_size": 0.144,
      "y_size": 1.218,
      "z_size": 2.026
    },
    {
      "object": "curtain",
      "x": -2.067,
      "y": 0.687,
      "z": -0.087,
      "x_size": 0.143,
      "y_size": 1.352,
      "z_size": 2.349
    },
    {
      "object": "chair",
      "x": -1.764,
      "y": -0.945,
      "z": -0.83,
      "x_size": 0.669,
      "y_size": 0.581,
      "z_size": 0.859
    },
    {
      "object": "chair",
      "x": -0.704,
      "y": -0.954,
      "z": -0.767,
      "x_size": 0.511,
      "y_size": 0.505,
      "z_size": 0.688
    },
    {
      "object": "chair",
      "x": -0.617,
      "y": 0.842,
      "z": -0.694,
      "x_size": 0.515,
      "y_size": 0.689,
      "z_size": 0.528
    },
    {
      "object": "pillow",
      "x": -0.637,
      "y": 0.859,
      "z": -0.607,
      "x_size": 0.435,
      "y_size": 0.218,
      "z_size": 0.385
    },
    {
      "object": "pillow",
      "x": -0.605,
      "y": 0.705,
      "z": -0.629,
      "x_size": 0.52,
      "y_size": 0.18,
      "z_size": 0.403
    },
    {
      "object": "chair",
      "x": -1.733,
      "y": 1.12,
      "z": -0.808,
      "x_size": 0.647,
      "y_size": 0.587,
      "z_size": 0.816
    },
    {
      "object": "cabinet",
      "x": -1.439,
      "y": 1.507,
      "z": -0.247,
      "x_size": 1.216,
      "y_size": 0.413,
      "z_size": 2.01
    },
    {
      "object": "tv_monitor",
      "x": 0.659,
      "y": -1.363,
      "z": -0.345,
      "x_size": 1.01,
      "y_size": 0.352,
      "z_size": 0.699
    },
    {
      "object": "cabinet",
      "x": 0.56,
      "y": -1.276,
      "z": -0.919,
      "x_size": 0.827,
      "y_size": 0.519,
      "z_size": 0.673
    },
    {
      "object": "cabinet",
      "x": -0.205,
      "y": -1.336,
      "z": -0.984,
      "x_size": 0.5,
      "y_size": 0.384,
      "z_size": 0.518
    },
    {
      "object": "cabinet",
      "x": -1.496,
      "y": -1.366,
      "z": 0.439,
      "x_size": 1.375,
      "y_size": 0.347,
      "z_size": 1.634
    },
    {
      "object": "bed",
      "x": 1.027,
      "y": 1.014,
      "z": -0.707,
      "x_size": 2.3,
      "y_size": 1.438,
      "z_size": 1.15
    },
    {
      "object": "pillow",
      "x": 1.784,
      "y": 0.654,
      "z": -0.601,
      "x_size": 0.406,
      "y_size": 0.619,
      "z_size": 0.087
    },
    {
      "object": "cabinet",
      "x": 1.716,
      "y": 0.104,
      "z": -0.944,
      "x_size": 0.766,
      "y_size": 0.561,
      "z_size": 0.627
    },
    {
      "object": "tv_monitor",
      "x": -0.431,
      "y": 1.592,
      "z": -0.349,
      "x_size": 0.505,
      "y_size": 0.245,
      "z_size": 0.349
    },
    {
      "object": "table",
      "x": -0.474,
      "y": 1.46,
      "z": -0.87,
      "x_size": 0.85,
      "y_size": 0.546,
      "z_size": 0.768
    },
    {
      "object": "curtain",
      "x": -2.045,
      "y": -0.526,
      "z": -0.1,
      "x_size": 0.296,
      "y_size": 1.525,
      "z_size": 2.326
    },
    {
      "object": "pillow",
      "x": 1.556,
      "y": 1.417,
      "z": -0.499,
      "x_size": 0.66,
      "y_size": 0.444,
      "z_size": 0.184
    }
  ]
}
  
  ```
## ---> We can now start the first training.

**Comment by bachvudinh** [comment]
## Fix the training bug.
- Due to large room dimensions in the data, 2D scans generate high-resolution video sequences that exceed our training capacity. These sequences reach over 12,000 tokens, while our A6000 GPUs can only process sequences up to 4,096 tokens. To address this limitation, I've filtered out these oversized samples and reduced the minimum pixels per unit from 50 to 35 when generating training data.
- The Training loss go down smoothly to 0.23 at the end of epoch:

 ![Image](https://github.com/user-attachments/assets/bdcbef8c-c01d-495e-8fa6-ed043df9fd78)

## Today
- Finish the benchmarking script and Work on the augmentation data with @new5558 .

**Comment by bachvudinh** [comment]
## Updated :
- Training run with bounding box not working, likely due to Qwen2.5-vl limitations. We'll pivot to using static images as demonstrated in the VoxRep paper like this:
  
 ![Image](https://github.com/user-attachments/assets/babb1d1b-d195-4877-bbde-2105d5771199)
 
- Also Norapat has resolved the bounding box issue in the data pipeline
### Training Experiment details:

| Run ID | Date       | Model Config | Dataset | Learning Rate | Batch Size | Seq Length | Max Resolution | Loss | Hardware | 
|--------|------------|--------------|---------|---------------|------------|-------|------|----------|----------|
| exp-1 | 2025-04-8 | [Full-Finetune](https://github.com/janhq/visual-thinker/blob/main/training/Llama-factory-config/Qwen2.5_1.5B_distil.yaml) | 6k 2D Room Scan Video | 1e-4 | 32  | 4096 |   300x300 | 0.23 | ~ 1.5 hours on 6xA6000s | 
| exp-2 | 2025-04-9 | [Full-Finetune](https://github.com/janhq/visual-thinker/blob/main/training/Llama-factory-config/Qwen2.5_1.5B_distil.yaml) | 6k 2D Room Scan Video | 1e-4 | 32  | 8192 |   400x400 | 0.2 | ~ 20 hours on 6xA6000s | 
| exp-3 | 2025-04-11 | [Full-Finetune](https://github.com/janhq/visual-thinker/blob/main/training/Llama-factory-config/Qwen2.5_1.5B_distil.yaml) | 6k 2D Room Scan Video without bounding box | 3e-4 | 32  | 8192 |   400x400 | 0.06 | ~ 20 hours on 6xA6000s | 
| exp-4 | 2025-04-11 | [Full-Finetune](https://github.com/janhq/visual-thinker/blob/main/training/Llama-factory-config/Qwen2.5_1.5B_distil.yaml) | Scannet Subset  | 1e-4 | 32  | 8192 |   400x400 |  | ~  on 8xH100 | 
| exp-5 | 2025-04-16 | [Full-Finetune](https://github.com/janhq/visual-thinker/blob/main/training/Llama-factory-config/Qwen2.5_1.5B_distil.yaml) | 26k 2D Room Scan Images without bounding box  | 1e-4 | 64  | 4096 |   1280x1280 | 0.02 | ~14  on 8xA6000 | 
| exp-6 | 2025-04-16 | [Full-Finetune](https://github.com/janhq/visual-thinker/blob/main/training/Llama-factory-config/Qwen2.5_1.5B_distil.yaml) | 26k 2D Room Scan Images without bounding box  | 1e-4 | 64  | 4096 |   896x896 | 0.01 | ~4  on 8xA6000 | 

### Today:
- Start training run with images data with Qwen-2.5Vl and Gemma3-4B[Finished]

**Comment by bachvudinh** [comment]
# Summary on next strategy:
## Idea end-to-end training approach using the pre-trained 3D-JEPA encoder: 
### Overall
- Meta's Locate 3D advances 3D object localization with 3D-JEPA encoder. Current limitations: only outputs bounding boxes for specific expressions
- Our proposal: 
  + fuse 3D-JEPA with LLM/VLM in end-to-end model
  + Create "bridge" translating spatial features to LLM-compatible tokens
  + Joint fine-tuning to enable 3D scene captioning and visual QA
- Goal: broader language-grounded 3D scene understanding
### Detailed Plan
The core idea is to treat the output of the 3D-JEPA encoder as a specialized "visual" modality that the LLM/VLM needs to understand and reason about alongside text:
- Input: Posed RGB-D Sequence.
- Preprocessing (Locate 3D Phase 1):
  + Lift 2D features (CLIP, DINOv2) from RGB frames onto a 3D point cloud using camera poses and depth.
  Voxelize/aggregate features.
  + Result: Featurized Point Cloud (PtClift) = {(x_i, y_i, z_i, f_i)} where f_i is the rich lifted feature vector for point/voxel i.
- Using Pretrained 3d-Jepa Encoder:
  + Input: PtClift
  + Output: A set of contextualized point/voxel feature vectors. Let's say N output points/tokens, each with dimension D_jepa. Result: JEPA_Features = [feat_1, feat_2, ..., feat_N] (where feat_j is dimension D_jepa). Note: These features capture the context of the entire scene.
- Convert the set JEPA_Features into a sequence of tokens (embeddings) that the LLM/VLM can process, similar to how image patches or text words are tokenized. Let the LLM/VLM's expected embedding dimension be D_llm.
- Input Construction: Create the input sequence for the LLM/VLM by concatenating:
  (For Captioning): [Scene_Tokens] + [Prompt Tokens (e.g., "Describe this scene:") or Task Token]
  (For VQA): [Scene_Tokens] + [Question Tokens] + [Prompt Tokens (e.g., "Answer:")]
- Training Strategy:
    + Stage 1 (Adapter/Bridge Training): Freeze both the 3D-JEPA encoder and the LLM/VLM. Train only the bridging layer (projection + queries) for a few epochs. This lets the bridge learn to translate 3D features into the LLM's space without disrupting the base models.
    + Stage 2 (Full Fine-tuning): Unfreeze all components (or keep the 3D-JEPA encoder frozen longer if needed) and fine-tune end-to-end.

**Comment by new5558** [comment]
## Research Field: 3D Semantic Scene Completion (SSC) 

![Image](https://github.com/user-attachments/assets/9966091c-a25f-4d23-b3ac-71d3bbd637a8)

![Image](https://github.com/user-attachments/assets/c4651ea3-b460-454d-80c4-57304e9483ab)

Objective: Given 1 frame RGB-D image, predict the entire scene in voxel (or other representation) space
Milestone work: https://arxiv.org/pdf/1611.08974 (2017)

**Comment by new5558** [comment]
### Current SOTA for 3d scene completion for indoor envs: **SG-SSC**
Link: https://github.com/aipixel/SG-SSC

![Image](https://github.com/user-attachments/assets/a3d726ed-416c-439a-a597-5962147a4077)

Findings from playing their code:
1. generated voxel space is fixed; no matter the room size is
2. result is not recognizable, this one may be bug

**Comment by new5558** [comment]
### Proposed idea: Train a model for bounding box understanding with a single-frame RGB-D image

![Image](https://github.com/user-attachments/assets/a3c1433b-07cf-407a-bd0d-6dc18545de2e)

**Input:** single frame RGB-D image
**output:** same as Locate 3D output
**backbone:** Locate 3D pipeline + 3D Scene completion model
**Dataset:** May need to use [SUNRGB-D 3D](https://rgbd.cs.princeton.edu/) dataset because 

#### Challenges
1. Locate 3D input is not actually **a voxel** but **a voxel with CLIP feature** <- need modification
2. Need to evaluate the **performance of 3D scene completion model** more

**Comment by new5558** [comment]
Closed because the issue is now seperated into
#4 voxel 3d object detection dataset
#5 voxel 3d object detection training
#6 next research direction based on meta locate 3d paper
